### PR TITLE
Fix disable shape inference for optimization

### LINKED ISSUE #649: "disable_shape_inference" in ORTOptimizer does not work
### System Info

```shell
optimum==1.6.1
```

### Who can help?

@echarlaix @philschmid 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Using following opt. config:
```
opt_conf = OptimizationConfig(optimization_level=0)
opt_conf.fp16 = True
opt_conf.enable_transformers_specific_optimizations = False
opt_conf.disable_embed_layer_norm_fusion = False
opt_conf.disable_shape_inference = True
```
The error keeps appearing (GPT-J model): `ZeroDivisionError: integer division or modulo by zero`

Error trace contains this:
```
File ~/miniconda3/envs/pytorch/lib/python3.10/site-packages/onnxruntime/transformers/models/gpt2/../../onnx_model.py:594, in OnnxModel.convert_float_to_float16(self, use_symbolic_shape_infer, **kwargs)
    591 if use_symbolic_shape_infer:
    592     # Use symbolic shape inference since custom operators (like Gelu, SkipLayerNormalization etc) are not recognized by onnx shape inference.
    593     shape_infer_helper = SymbolicShapeInferenceHelper(model)
--> 594     model = shape_infer_helper.infer_shapes(model, auto_merge=True, guess_output_rank=False)
    596 parameters = {"disable_shape_infer": use_symbolic_shape_infer}
    597 parameters.update(
    598     {
    599         key: kwargs[key]
   (...)
    609     }
    610 )
```

And if you check the code in optimum/onnxruntime/optimization.py on line number 159 you will find that
`optimizer.convert_float_to_float16(keep_io_types=True)` doesn't get a parameter that disables symbolic shape inference.

However, the signature of this method is:
`def convert_float_to_float16(self, use_symbolic_shape_infer=True, **kwargs):`

### Expected behavior

I would like the errors with shape inference to be better explained and to be able to turn it off for real.

### PR DESCRIPTION
**Pull request by regisss**
# What does this PR do?

Fixes #649. The `disable_shape_inference` field of the optimization configuration should be taken into account when calling `convert_flot_to_fp16` while optimizing a model.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/onnxruntime/optimization.py [hunk 1]
```diff
@@ -143,20 +143,30 @@ def optimize(
...
+                    # keep_io_types to keep inputs/outputs as float32
+                    optimizer.convert_float_to_float16(
+                        use_symbolic_shape_infer=not optimization_config.disable_shape_inference, keep_io_types=True
+                    )
+            except Exception as e:
+                if "Incomplete symbolic shape inference" in str(e):
+                    err = RuntimeError(
+                        f"{str(e)}. Try to set `disable_shape_inference=True` in your optimization configuration."
+                    )
+                    raise err from e
```
Comment (fxmarty) [inline optimum/onnxruntime/optimization.py:168]: How does it compare to:
```
raise RuntimeError(f"{str(e)}. Try to set `disable_shape_inference=True` in your optimization configuration.")
```

?

Comment (regisss) [reply optimum/onnxruntime/optimization.py:168]: The current solution returns:
```
Traceback (most recent call last):
  File "/home/regis/HuggingFace/dev/optimum/optimum/onnxruntime/optimization.py", line 160, in optimize
    optimizer.convert_float_to_float16(
  File "/home/regis/HuggingFace/dev/venv/lib/python3.8/site-packages/onnxruntime/transformers/models/gpt2/../../onnx_model.py", line 594, in convert_float_to_float16
    model = shape_infer_helper.infer_shapes(model, auto_merge=True, guess_output_rank=False)
  File "/home/regis/HuggingFace/dev/venv/lib/python3.8/site-packages/onnxruntime/transformers/models/gpt2/../../../tools/symbolic_shape_infer.py", line 2347, in infer_shapes
    raise Exception("Incomplete symbolic shape inference")
Exception: Incomplete symbolic shape inference

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test_optimum.py", line 26, in <module>
    optimizer.optimize(save_dir=onnx_path, optimization_config=optimization_config)
  File "/home/regis/HuggingFace/dev/optimum/optimum/onnxruntime/optimization.py", line 168, in optimize
    raise err from e
RuntimeError: Incomplete symbolic shape inference. Try to set `disable_shape_inference=True` in your optimization configuration.
```

While raising directly the error returns:
```
Traceback (most recent call last):
  File "/home/regis/HuggingFace/dev/optimum/optimum/onnxruntime/optimization.py", line 160, in optimize
    optimizer.convert_float_to_float16(
  File "/home/regis/HuggingFace/dev/venv/lib/python3.8/site-packages/onnxruntime/transformers/models/gpt2/../../onnx_model.py", line 594, in convert_float_to_float16
    model = shape_infer_helper.infer_shapes(model, auto_merge=True, guess_output_rank=False)
  File "/home/regis/HuggingFace/dev/venv/lib/python3.8/site-packages/onnxruntime/transformers/models/gpt2/../../../tools/symbolic_shape_infer.py", line 2347, in infer_shapes
    raise Exception("Incomplete symbolic shape inference")
Exception: Incomplete symbolic shape inference

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test_optimum.py", line 26, in <module>
    optimizer.optimize(save_dir=onnx_path, optimization_config=optimization_config)
  File "/home/regis/HuggingFace/dev/optimum/optimum/onnxruntime/optimization.py", line 170, in optimize
    raise e
RuntimeError: Incomplete symbolic shape inference. Try to set `disable_shape_inference=True` in your optimization configuration.
```

I think the message `The above exception was the direct cause of the following exception:` makes things less confusing.

I also thought about directly modifying the message of the exception raised by ORT but I think it's better to have a clear separation between ORT and Optimum error messages.

Comment (fxmarty) [reply optimum/onnxruntime/optimization.py:168]: Oh I see, yes agreed!

**Review by fxmarty — APPROVED** [review]
LGTM!

**Pull request merged**